### PR TITLE
Add result type inference to op definition

### PR DESCRIPTION
**Pull request by tungld**
There are many ONNX operators that can be expressed by other ONNX operators. For example, ONNXReduceSumSquareOp is composition of ONNXReduceSumOp and ONNXMulOp, like
```
ONNXReduceSumSquareOp %X = ONNXReduceSumOp (ONNXMulOp %X, %X)
```

Currently, we did the above rewriting by sub-classing `RewritePattern`, it means we write it manually. 

However, it is better to express it by using declarative rewriting rules like:
```
def ReduceSumSquareOpPattern: Pat<(ONNXReduceSumSquareOp $oprd, $axes, $keepdims),
       (ONNXReduceSumOp (ONNXMulOp $oprd, $oprd), $axes, $keepdims)>;
```
To enable the above declarative rewriting rules, the operator `ONNXMulOp`'s build methods must support result type deduction. In order to do so, we add custom build methods into the Op definition. These custom build methods have similar signatures as auto-generated build methods, except the absence of result type which will be deduced. 

For most of the cases, the result type can be a `UnrankedTensorType` whose element type is the same as the first operand's element type, and we let shape inference infer the final result type. Below is one example of custom build methods for ONNXMulOp:
```tablegen
let builders = [
     OpBuilder<"Builder *builder, OperationState &state,Value A, Value B", [{
        auto elementType = A.getType().cast<TensorType>().getElementType();
        build(builder, state, UnrankedTensorType::get(elementType), A, B);
     }]>
];
```

This patch is to realize the above idea of supporting declarative rewriting rules.

In order to auto-generate custom build methods for a Op, we add the Op in the the list `custom_builder_ops_list` in `gen_doc.py`.

### CODE REVIEW

**Comment by tungld** [comment]
I am investigating why `check-mlir-lit` failed in the buildbot, while it ran successfully in my local environment.

File: src/dialect/onnx/onnx_ops.cpp [hunk 1]
```diff
@@ -272,6 +283,16 @@ void ONNXMulOp::inferShapes() {
...
+/// Infer the result type of the ONNXMulOp. This method is required by the inter
+/// type op interace.
+LogicalResult ONNXMulOp::inferReturnTypes(
+    MLIRContext *, Optional<Location> location, ValueRange operands,
+    ArrayRef<NamedAttribute> attributes, RegionRange regions,
+    SmallVectorImpl<Type> &inferedReturnTypes) {
+  inferedReturnTypes.assign({operands[0].getType()});
+  return success();
+}
+
```
Comment (chentong319) [inline src/dialect/onnx/onnx_ops.cpp]: This code can be automated also. For most of Op, the output type is the same as one of the input. This rule is specified by onnx. If you'd like, we can discuss this issue at the meeting next week.

Comment (tungld) [reply src/dialect/onnx/onnx_ops.cpp]: You are right. In the latest code, I just return an UnrankedTensorType with the same element type as the first input. Is it what you suggested? It's very nice to discuss this issue at the next meeting.

**Comment by tungld** [comment]
> This patch solved some special cases in which an Op is a composite of other Ops. Is that true? Can we decompose the Op into sub ops to get the same effect?

Yes, it's true. The main purpose is to decompose an Op into `primitive` Ops. In that way, we only need to manually lower the primitive Ops.

**Comment by tungld** [comment]
It seems that the buildbot is using the latest code of llvm-project where `InferTypeOpInterface` has a verifier to check the result type (c.f. https://github.com/llvm/llvm-project/blob/master/mlir/include/mlir/Analysis/InferTypeOpInterface.td#L64). Verification failed for an Op which has both shape inference and result type inference, since the shape inference produced a RankedTensorType while the result type inference produced a UnrankedTensorType.

I wonder whether I can override `InferTypeOpInterface` in order to disable verifier.

**Comment by tungld** [comment]
@tjingrant It seems the buildbot triggered installing the latest MLIR and it failed in installing MLIR.

**Comment by chentong319** [comment]
I tried a test for PadConstantValue pattern with unknown rank.
// CHECK-LABEL: @test_constant_pad(%{{.*}}: tensor<?x?xf32>) -> tensor<?xf32> {
func @test_constant_pad(%arg0 : tensor<?x?xf32>) -> tensor<?xf32> {
  // CHECK-NEXT: [[SQUARE:%.+]] = "onnx.PadConstantValuePad"(%arg0) {constant_value = 0.000000e+00 : f32, mode = "constant", pads = [0, 2, 0, 0]} : (tensor<?x?xf32>) -> tensor<?xf32>
  %0 ="onnx.Constant"() {value=[0, 2, 0, 0]} : ()-> tensor<?xi64>
  %2 = "onnx.PadConstantValue"(%arg0, %0) {constant_value=0. : f32, mode = "constant"} : (tensor<?x?xf32>, tensor<?xi64>)-> tensor<?xf32>
  "std.return"(%2) : (tensor<?xf32>) -> ()
}

It works with the pattern: (onnx_combined.td in the master)
def ConstantPadPattern : Pat<(ONNXPadConstantValueOp $m1, (ONNXConstantOp:$res $v1, $v2), $m2, $m3),
                             (ONNXPadConstantValuePadOp $m1, $v2, $m2, $m3),
                             [(HasOneUse $res)]>;

Is your pattern similar to this pattern?

**Comment by tungld** [comment]
> def ConstantPadPattern : Pat<(ONNXPadConstantValueOp $m1, (ONNXConstantOp:$res $v1, $v2), $m2, $m3),
> (ONNXPadConstantValuePadOp $m1, $v2, $m2, $m3),
> [(HasOneUse $res)]>;

This is not the case in this patch, since the target pattern has only one op, `ONNXPadConstantValuePadOp`, and to construct it, the result type is getting from the result type of `ONNXPadConstantValueOp` in the source pattern.

If the target pattern is constructed by two or more ops, e.g.: `ONNXPadConstantValuePadOp $m1, $v2, (ONNXExpOp $m2), (ONNXNegOp $m3))`, then it is the case for this patch.

**Comment by chentong319** [comment]
@tungld I see. It looks like more a MLIR issue.

**Pull request merged**